### ISSUE TITLE
FFT error

### ISSUE THREAD

**Issue by ViralBShah** [JuliaComputing/ArrayFire.jl#36]
````
julia> a = rand(Float32, 100,100);

julia> b = AFArray(a);

julia> fft(a)-fft(b);
ERROR: BoundsError: attempt to access (100,100)
  at index [3]
 in - at arraymath.jl:97
````

**Comment by ViralBShah** [comment]
Perhaps we need some promotion rules - or at least better error messages, if mixed CPU/GPU operations are not supported.

**Comment by ranjanan** [comment]
You cannot subtract an `Array` and an `AFArray`. I could of course change the default behavior to transferring to the CPU and then subtracting.

**Comment by ranjanan** [comment]
Yeah, the error messages could be better I think. I might simply say 
`-(a::Array, b::AFArray) = throw("Can't subtract Arrays and AFArrays")`

**Comment by ViralBShah** [comment]
Capture in docs? Ideally, if someone tries to do these mixed operations, they should get a useful error.  It is not enough to add just that one line. You'll need to add this comprehensively for all operations in all positions.

**Comment by ViralBShah** [comment]
I say let's just have a section in the docs in the compute model - that for purposes of performance, we do not automatically move data and it will be an error.

**Comment by ranjanan** [comment]
Correct, I'd have to do this for all functions. I think I can capture this in the docs.

**Comment by ViralBShah** [comment]
That's quite a bit of difference, but perhaps one can't do much about it. And I am not even running on a real GPU - just on my mac.

````
julia> maximum(abs(fft(a)-Array(fft(b))))
0.00029078507f0
````

**Comment by ranjanan** [comment]
If you're using the OpenCL backend, that probably has to do with the different in the value returned by clFFT and FFTW. I'm not sure what we can do about those numerical differences. They would also arise in case you're using CuFFT on a GPU too.

**Comment by ranjanan** [comment]
Do you think it's worth noting in the README about the differences in the values?

**Comment by SimonDanisch** [comment]
clFFT (which is probably also their CPU backend) is quite likely the culprit here ;)
I coincidentally just tried out clFFT directly and got a very similar result:
`maximum(abs(a - b)) ->  0.000159`

**Comment by SimonDanisch** [comment]
CUDA backend seems to be in the same range:
```Julia
julia> maximum(abs(af-Array(bf)))
0.0001501969f0
```

**Comment by ViralBShah** [comment]
Perhaps just one line to say that the numerical results may differ from the CPU versions that may sometimes be more accurate. Users should adequately check their programs for correctness, as they always should.

**Comment by ranjanan** [comment]
I've added notes on both.

**Comment by alanedelman** [comment]
how do you know the cpu is more accurate?
random rounding errors are random rounding errors
if your code depends on this something is wrong

**Comment by pavanky** [comment]
FFTs can be notoriously hard to verify numerically because

1) floating point arithmetic is not associative
2) it uses a lot of exponential operations.

Any numerical changes caused by (1) will result in larger changes by (2).

**Comment by alanedelman** [comment]
Your assertions 1 and 2 are true, your conclusion is not.

**Comment by pavanky** [comment]
Welp. should have worded it differently.

**Comment by alanedelman** [comment]
fft's are not any more difficult to verify numerically than anything else

**Comment by pavanky** [comment]
I mean't hard to verify for correctness in the way @ViralBShah was doing. 

Anyway according the published specs, the accuracy for expoentials seems to be the following:
- 1ULP for x86 (http://www.gnu.org/software/libc/manual/html_node/Errors-in-Math-Functions.html)
- 2ULP for CUDA (http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#mathematical-functions-appendix)
- 3ULP for OpenCL (https://www.khronos.org/registry/cl/specs/opencl-2.1-environment.pdf)

**Comment by ViralBShah** [comment]
I was generally thinking like many users probably will - that CPU libraries tend to have slightly higher accuracy with ULPs and such and accelerated libraries will give up some last bits to be faster. Also @stevengj 's FFTW is a pretty good reference, I would have thought.
